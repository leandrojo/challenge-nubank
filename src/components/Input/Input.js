import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import Masker from 'vanilla-masker';

const StyledInput = styled.input`
  appearance: none;
  border: none;
  box-shadow: none;
  width: 100%;

  ${p => p.theme.components.input.css};

  &:focus {
    border-color: ${props => props.theme.colors.secondary};
    outline: none;
  }
`;

const ref = React.createRef();

export const Input = ({
  errors,
  id,
  isMoney,
  label,
  moneyFormatter,
  name,
  onChange,
  pattern,
  value,
  ...rest
}) => {
  const [state, setState] = useState({
    value: '',
  });

  function unformat(v) {
    if (isMoney) {
      return parseFloat(Masker.toNumber(v));
    }

    return v;
  }

  function getValue() {
    return state.value;
  }

  useEffect(() => {
    function format(unformatted) {
      if (isMoney) {
        return Masker.toMoney(Masker.toNumber(unformatted), moneyFormatter);
      }

      if (typeof unformatted === 'string' && pattern !== '') {
        return Masker.toPattern(unformatted, pattern);
      }

      return unformatted;
    }

    setState({
      value: format(value),
    });
  }, [isMoney, moneyFormatter, pattern, value]);

  const handleBlur = () => {};

  const handleFocus = () => {};

  const handleChange = (ev) => {
    onChange(unformat(ev.target.value), name);
  };

  return (
    <StyledInput
      id={id}
      name={name}
      onBlur={handleBlur}
      onChange={handleChange}
      onFocus={handleFocus}
      pattern="\d*"
      ref={ref}
      type="text"
      value={getValue()}
      {...rest}
    />
  );
};

const moneyFormatter = {
  precision: 2,
  separator: '.',
  delimiter: ',',
  unit: 'US$',
};

Input.defaultProps = {
  errors: [],
  id: '',
  isMoney: false,
  label: '',
  moneyFormatter,
  onChange: () => {},
  onError: () => {},
  pattern: '',
  rules: [],
};

const RulePropTypes = PropTypes.oneOfType([PropTypes.string, PropTypes.func]);
const MoneyFormatterPropTypes = PropTypes.shape({
  precision: PropTypes.number,
  separator: PropTypes.string,
  delimiter: PropTypes.string,
  unit: PropTypes.string,
});

Input.propTypes = {
  errors: PropTypes.arrayOf(PropTypes.string),
  id: PropTypes.string,
  isMoney: PropTypes.bool,
  label: PropTypes.string,
  pattern: PropTypes.string,
  moneyFormatter: MoneyFormatterPropTypes,
  name: PropTypes.string.isRequired,
  onError: PropTypes.func,
  onChange: PropTypes.func,
  rules: PropTypes.arrayOf(RulePropTypes),
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
};

export default Input;
