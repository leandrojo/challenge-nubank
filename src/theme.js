const fontFamily = 'Rubik, sans-serif';

const theme = {
  backgroundColor: '#f5f5f5',
  colors: {
    primary: '#7f39b7',
    grayExtraDark: '#555',
    grayDark: '#767676',
    gray: '#999',
    grayLight: '#ccc',
    grayExtraLight: '#FAFAFA',
  },
  fontFamily,
  fontSize: '1em',
  typography: {
    heading: {
      color: '#252226',
    },
    paragraph: {
      color: '#716d73',
    },
  },
};

theme.components = {
  input: {
    css: {
      backgroundColor: 'transparent',
      fontFamily,
      fontSize: '1.4em',
      fontWeight: 300,
    },
  },
  pane: {
    backgroundColor: '#fff',
  },
  range: {
    emptyColor: theme.colors.grayLight,
  },
};

export default theme;
