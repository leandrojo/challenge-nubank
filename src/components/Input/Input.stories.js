/* eslint-disable import/no-extraneous-dependencies */

import React, { useState } from 'react';

import Input from './Input';

export default {
  title: 'components/Input',
};

export const custom = () => <Input placeholder="Custom Placeholder" />;

custom.story = {
  name: 'default custom',
};

export const money = () => {
  const [value, setValue] = useState('1000.00');
  return <Input isMoney onChange={setValue} value={value} />;
};

money.story = {
  name: 'money',
};

const moneyFormatter = {
  precision: 2,
  separator: ',',
  delimiter: '.',
  unit: 'R$',
};

export const BRL = () => {
  const [value, setValue] = useState('1000,00');
  return <Input isMoney onChange={setValue} value={value} moneyFormatter={moneyFormatter} />;
};

BRL.story = {
  name: 'money formatter',
};
