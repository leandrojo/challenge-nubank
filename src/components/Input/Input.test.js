/* eslint-env jest */

import React, { useState } from 'react';
import { ThemeProvider } from 'styled-components';
import renderer from 'react-test-renderer';

import theme from 'theme';

import Input from './Input';

describe('Input', () => {
  it('with initial value', () => {
    const Component = () => {
      const [value, onChange] = useState(1000);

      return (
        <ThemeProvider theme={theme}>
          <Input name="firstDeposit" onChange={onChange} value={value} />
        </ThemeProvider>
      );
    };

    const expected = renderer.create(<Component />).toJSON();

    expect(expected).toMatchSnapshot();
  });
});
