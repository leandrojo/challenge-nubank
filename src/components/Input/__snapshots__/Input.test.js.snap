// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input with initial value 1`] = `
<input
  className="sc-bdVaJa jIXIXs"
  id=""
  name="firstDeposit"
  onBlur={[Function]}
  onChange={[Function]}
  onError={[Function]}
  onFocus={[Function]}
  pattern="\\\\d*"
  type="text"
  value=""
/>
`;
